src/binned_allocator.zig

```diff
@@ -0,0 +1,527 @@
+            } else {
+                return null;
+            }
+        }
+
+        fn resize(ctx: *anyopaque, buf: []u8, log2_align: u8, new_len: usize, ret_addr: usize) bool {
+            const self: *Self = @ptrCast(@alignCast(ctx));
+
+            const align_ = @as(usize, 1) << @as(std.math.Log2Int(usize), @intCast(log2_align));
+            comptime var prev_size: usize = 0;
+            inline for (&self.bins) |*bin| {
+                if (buf.len <= bin.size and align_ <= bin.size) {
+                    // Check it still fits
+                    return new_len > prev_size and new_len <= bin.size;
+                }
+                prev_size = bin.size;
+            }
+
+            // Assuming it's a large alloc
+            if (new_len <= prev_size) return false; // New size fits into a bin
+            return self.backing_allocator.rawResize(buf, log2_align, new_len, ret_addr);
+        }
+
+        fn free(ctx: *anyopaque, buf: []u8, log2_align: u8, ret_addr: usize) void {
+            const self: *Self = @ptrCast(@alignCast(ctx));
+
+            const align_ = @as(usize, 1) << @as(std.math.Log2Int(usize), @intCast(log2_align));
+            inline for (&self.bins) |*bin| {
+                if (buf.len <= bin.size and align_ <= bin.size) {
+                    bin.free(buf.ptr);
+                    return;
+                }
+            }
+
+            // Assuming it's a large alloc
+            self.backing_allocator.rawFree(buf, log2_align, ret_addr);
+            if (config.report_leaks) {
+                self.large_count.decrement();
+            }
+        }
+
+        const Mutex = if (config.thread_safe)
+            std.Thread.Mutex
+        else
+            struct {
+                fn lock(_: @This()) void {}
+                fn unlock(_: @This()) void {}
+            };
+
+        const Counter = if (config.thread_safe)
+            struct {
+                count: std.atomic.Atomic(usize),
+                fn init() @This() {
+                    return .{ .count = std.atomic.Atomic(usize).init(0) };
+                }
+                fn load(self: @This()) usize {
+                    return self.count.load(.Acquire);
+                }
+                fn increment(self: @This()) void {
+                    _ = self.count.fetchAdd(1, .AcqRel);
+                }
+                fn decrement(self: @This()) void {
+                    _ = self.count.fetchSub(1, .AcqRel);
+                }
+            }
+        else
+            struct {
+                count: usize,
+                fn init() @This() {
+                    return .{ .count = 0 };
+                }
+                fn load(self: @This()) usize {
+                    return self.count;
+                }
+                fn increment(self: @This()) void {
+                    self.count += 1;
+                }
+                fn decrement(self: @This()) void {
+                    self.count -= 1;
+                }
+            };
```

Comment:
```suggestion
        const Counter = if (config.thread_safe)
            struct {
                count: std.atomic.Atomic(usize),
                fn init() @This() {
                    return .{ .count = std.atomic.Atomic(usize).init(0) };
                }
                fn load(self: *const @This()) usize {
                    return self.count.load(.Acquire);
                }
                fn increment(self: *@This()) void {
                    _ = self.count.fetchAdd(1, .AcqRel);
                }
                fn decrement(self: *@This()) void {
                    _ = self.count.fetchSub(1, .AcqRel);
                }
            }
        else
            struct {
                count: usize,
                fn init() @This() {
                    return .{ .count = 0 };
                }
                fn load(self: @This()) usize {
                    return self.count;
                }
                fn increment(self: *@This()) void {
                    self.count += 1;
                }
                fn decrement(self: *@This()) void {
                    self.count -= 1;
                }
            };
```
I believe the `*const @This()` is necessary for the atomic `load` function to ensure correct semantics (avoid a non-atomic copy of `count` when passing the argument).